lib/gpfpiso.gi

```diff
@@ -1851,8 +2053,120 @@ local isob,isos,iso,gens,u,a,rels,l,i,j,bgens,cb,cs,b,f,k,w,monoid,
+#local f,miso,lm,mg,l,pre,gens,m,bas,pownum;
+#  pownum:=function(w)
+#  local l,s,i;
+#    l:=LetterRepAssocWord(w);
+#    s:=0;
+#    for i in [1..Length(l)] do
+#      s:=s+l[i]*bas^(i-1);
+#    od;
+#    return s;
+#  end;
+#  pre:=DataAboutSimpleGroup(Source(hom));
+#  pre:=ShallowCopy(pre.idSimple);
+#  if not IsBound(pre.parameter) then pre.parameter:=pre.shortname;fi;
+#  f:=Range(hom);
+#  miso:=IsomorphismFpMonoid(f);
+#  gens:=MappingGeneratorsImages(hom)[2];
+#  m:=Range(miso);
+#  mg:=List(GeneratorsOfMonoid(m));
+#  bas:=Length(mg);
+#  l:=[Size(f),pre.series,pre.parameter];
+#  if IsBound(pre.shortname) then Add(l,pre.shortname);fi;
+#  l:=[l,MappingGeneratorsImages(hom)[1]];
+#  Add(l,bas);
+#  bas:=bas+1; # need to go one larger, to avoid missing trailing 1's
+#  Add(l,List(mg,String));
+#  if List(mg,x->LetterRepAssocWord(UnderlyingElement(x)))<>
+#    List([1..Length(mg)],x->[x]) then Error("gens!"); fi;
+#  pre:=List(mg,x->LetterRepAssocWord(UnderlyingElement(
+#    PreImagesRepresentative(miso,x))));
+#  if ForAny(pre,x->Length(x)<>1) then Error("double");fi;
+#  Add(l,Concatenation(pre));
+#  if IsBound(m!.rewritingSystem) then
+#    mg:=m!.rewritingSystem;
+#    Add(l,LevelsOfGenerators(mg!.ordering));
+#  else
+#    Add(l,0);
+#  fi;
+#  Add(l,List(RelationsOfFpMonoid(m),x->List(x,pownum)));
+#  return l;
+#end;
+
+BindGlobal("BuildRewritingFromData",function(dat)
+local bas,fm,fam,mfam,mgens,gens,mrels,rels,i,f,g,m,hom;
+  bas:=dat[3]+1;
+  fm:=FreeMonoid(dat[4]);
+  mfam:=FamilyObj(One(fm));
+
+  f:=Filtered([1..Length(dat[5])],x->dat[5][x]>0);
+  f:=FreeGroup(dat[4]{f});
+  fam:=FamilyObj(One(f));
+
+  mrels:=[];
+  rels:=[];
+  for i in dat[7] do
+    i:=List(i,x->CoefficientsQadic(x,bas));
+    Add(mrels,List(i,x->AssocWordByLetterRep(mfam,x)));
+    i:=List(i,x->dat[5]{x});
+    # avoid the inverse relations
+    if not (Length(i[1])=2 and i[1][1]=-i[1][2]) then
+      i:=List(i,x->AssocWordByLetterRep(fam,x));
+      Add(rels,i[1]/i[2]);
+    fi;
+  od;
+  m:=fm/mrels;
+  if IsList(dat[6]) then
+    i:=KnuthBendixRewritingSystem(m,WreathProductOrdering(fm,dat[6]):isconfluent);
+    i!.reduced:=true;
+    m!.rewritingSystem:=i;
+  fi;
+
+  g:=f/rels;
+  mgens:=GeneratorsOfMonoid(fm);
+  gens:=GeneratorsOfGroup(f);
+
+  MakeFpGroupToMonoidHomType1(g,m);
+
+  hom:=GroupHomomorphismByImagesNC(Group(dat[2]),g,dat[2],GeneratorsOfGroup(g));
+  SetIsomorphismFpGroup(Source(hom),hom);
+  return hom;
+end);
+
+
+InstallMethod(IsomorphismFpGroupForRewriting,"simple groups: stored",
+  [IsSimpleGroup and IsFinite],1,
+function(G)
+local a,f,iso;
+
+  a:=DataAboutSimpleGroup(G);
+  if not IsBound(a.classicalId) then
+    TryNextMethod();
+  fi;
+  iso:=ShallowCopy(a.idSimple);
+  if not IsBound(iso.parameter) then
+    iso.parameter:=iso.shortname;
+  fi;
+
+  # is it a pre-stored presentation?
+  f:=First(List(GAPInfo.RootPaths,x->Concatenation(x,"grp/simplerew.grp")),
+       IsExistingFile);
```

Comment:
```suggestion
  f:=Filename(DirectoriesLibrary("grp"), "simplerew.grp");
```

Reply:
Food for thought, not meant to hold up the PR:

It seems wasteful to parse this file every time you use it (which takes ~160ms on my 2018 MBP with very fast SSD), creating ~2MB of data, most of which is discarded again. Sure, 160ms is not much if you need to run computations that are much slower; but it adds up?

There are several ways to address this (as you are of course well aware):
- we could just always load the data at the start of GAP (but of course then we slow down GAP startup, and increase the base memory usage, for data the user might never use)
- we could load it the first time it is used, but "cache" it by storing it some variable (we'd still waste those 2 MB afterwards, if it is only used once, but this could be mitigated by using `InstallFlushableValue`)
- the data could be stored in multiple files: e.g. all data for each series in one file; or even one file per group: something like `simplerew-C-2-3.grp` for series S, parameters [2,3].
- on the long run, having a simple fast builtin database would be nice... I've been thinking about doing just such a thing in OSCAR for some time (but of course as always it's difficult to make something that fits everyone's needs...

Reply:
@fingolfin 
I've thought about the issue of (re)loading, but in the end decided on the current solution as being most simple (and thus least prone to errors or in need of testing) and as not hogging up memory in permanence. 

The reason for doing so is that the only place this is going to be used is in calculating 2-cohomology (and extensions) for groups with larger simple composition factors. The data base will reduce really long KB runs (sometimes hours, a week in the case of M22) to a lookup. Even if that takes 10 seconds, it will be magnitudes faster.
Thus if this data base is used more than once, it will be in an outer loop iterating over many groups. For each of these groups the cohomology calculation still needs to iterate over all pairs of rules. This will take in the magnitude of minutes or more. A loading time of seconds will not make a meaningful difference.

There does not seem to be a "canned" solution (all data bases do their own thing), but of course if there was, this could easily be changed. (Though I don't think it would be a poster child for infrastructure either.)